### ISSUE TITLE
ALLEGRO_FULLSCREEN is broken on Linux

### ISSUE THREAD

**Issue by SiegeLord** [liballeg/allegro5#596]
E.g. on KWin it leaves some of the panels showing and is impossible to alt-tab out of.

**Comment by allefant** [comment]
I wonder if we should add a new ALLEGRO_FULLSCREEN_MODE (short for videoMODE change) which does what the current ALLEGRO_FULLSCREEN does, and change the meaning of ALLEGRO_FULLSCREEN to be either ALLEGRO_FULLSCREEN_WINDOW or ALLEGRO_FULLSCREEN_MODE. And the decision would depend on how much sense it makes - for example for iOS and Android there is no way to change the mode at all so it would be ALLEGRO_FULLSCREEN_WINDOW. And maybe the same on Linux. Independent of that it would be nice if we could fix ALLEGRO_FULLSCREEN(_MODE) for Linux of course.

**Comment by trentg** [comment]
Obviously consistency is important so it should be the same across all 
desktop OSes. I don't see a whole lot of point to the change though. 
First of all we have backwards compatibility to maintain and second, it 
saves maybe 1 line of code per game if your game runs on mobile and desktop.

**Comment by beoran** [comment]
We discussed something like this before, but in the end we didn't really get anywhere. Mode setting on Linux is a bit of a problem really, not in the least because some people who develop X11 are hostile to the concept. It looks like in the future, with Wayland, mode setting will be completely banned and a full screen window is all we will get...

**Comment by trentg** [comment]
Right, I forgot that some didn't want to support the features of the hardware but cripple it intentionally kind of like how they've done with some of the DEs. I still think as long as we support X we should support fullscreen. SDL doesn't have any problem doing so, so it's possible to do across all WMs.

**Comment by beoran** [comment]
Well, yes, SDL does support mode setting on X, so we should too, but I've both seen the code that SDL uses to do so and it isn't very beautiful, many hacks left and right are needed. So it's  pain to corect full screen mode setting on X if it doesn't work for some reason. Not to mention how the various window managers tend to interfere with proper operation of mode setting and full screen windows...

**Comment by snoopdouglas** [comment]
Does it still look like this has become infeasible? I wonder whether it'd just be worth either:

1. Just outright disabling `ALLEGRO_FULLSCREEN` for Linux and macOS in future releases (ref #566), and instead providing some simple example code showing how to draw to - and then scale up - a lower-res double buffer to fill the screen at desktop resolution

2. Magically scaling and drawing a fake backbuffer onto the real backbuffer to accomplish the latter part of (1) as part of `al_flip_display`.

I'd have a preference for (1) as it's less intrusive.
